{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "# API handling \n",
    "from easy_entrez import EntrezAPI\n",
    "# Parsing information\n",
    "from easy_entrez.parsing import xml_to_string\n",
    "\n",
    "import xml.etree.ElementTree as ET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<EntrezResponse status=200 for FetchQuery ['rs6311', 'rs662138'] in snp>\n",
      "<Element '{https://www.ncbi.nlm.nih.gov/SNP/docsum}DocumentSummary' at 0x0000022B786BEC00>\n"
     ]
    }
   ],
   "source": [
    "# Init class to use the API\n",
    "entrez_api = EntrezAPI(\n",
    "    'ret_toolkit',\n",
    "    'fernando.ggfigueroa@icloud.com',\n",
    "    return_type='json'\n",
    ")\n",
    "\n",
    "# Fetching genes for a variant from dbSNP\n",
    "ret_snp = entrez_api.fetch(['rs6311', 'rs662138'], max_results=2, database='snp')\n",
    "print(ret_snp)           # <EntrezResponse status=200 for FetchQuery ['rs6311'] in snp>\n",
    "print(ret_snp.data[0])   # <Element '{https://www.ncbi.nlm.nih.gov/SNP/docsum}DocumentSummary' at 0x108d22930>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\" ?>\n",
      "<ns0:DocumentSummary xmlns:ns0=\"https://www.ncbi.nlm.nih.gov/SNP/docsum\" uid=\"662138\">\n",
      "    <ns0:SNP_ID>662138</ns0:SNP_ID>\n",
      "    <ns0:ALLELE_ORIGIN/>\n",
      "    <ns0:GLOBAL_MAFS>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>1000Genomes</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.121487/608</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>ALSPAC</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.193046/744</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>Estonian</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.154464/692</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>GENOME_DK</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.225/9</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>GnomAD</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.138558/19420</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>GoNL</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.162325/162</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>HapMap</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.117647/192</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>KOREAN</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.001027/3</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>MGP</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.136704/73</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>NorthernSweden</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.155/93</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>PAGE_STUDY</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.117242/9226</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>PRJEB36033</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.088889/8</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>PRJEB37584</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.002525/2</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>PRJEB37766</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.291363/958</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>Qatari</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.097222/21</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>SGDP_PRJ</ns0:STUDY>\n",
      "            <ns0:FREQ>C=0.47541/58</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>Siberian</ns0:STUDY>\n",
      "            <ns0:FREQ>C=0.444444/8</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>TOPMED</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.141025/37328</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>TWINSUK</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.171521/636</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "        <ns0:MAF>\n",
      "            <ns0:STUDY>ALFA</ns0:STUDY>\n",
      "            <ns0:FREQ>G=0.160454/13541</ns0:FREQ>\n",
      "        </ns0:MAF>\n",
      "    </ns0:GLOBAL_MAFS>\n",
      "    <ns0:GLOBAL_POPULATION/>\n",
      "    <ns0:GLOBAL_SAMPLESIZE>0</ns0:GLOBAL_SAMPLESIZE>\n",
      "    <ns0:SUSPECTED/>\n",
      "    <ns0:CLINICAL_SIGNIFICANCE/>\n",
      "    <ns0:GENES>\n",
      "        <ns0:GENE_E>\n",
      "            <ns0:NAME>SLC22A1</ns0:NAME>\n",
      "            <ns0:GENE_ID>6580</ns0:GENE_ID>\n",
      "        </ns0:GENE_E>\n",
      "    </ns0:GENES>\n",
      "    <ns0:ACC>NC_000006.12</ns0:ACC>\n",
      "    <ns0:CHR>6</ns0:CHR>\n",
      "    <ns0:HANDLE>ACPOP,TISHKOFF,SI_EXO,EVA_UK10K_TWINSUK,CSHL-HAPMAP,KRGDB,1000GENOMES,EVA_DECODE,GNOMAD,GENOMED,EGCUT_WGS,1000G_HIGH_COVERAGE,GRF,HAMMER_LAB,BCMHGSC_JDW,BIOINF_KMB_FNS_UNIBA,PAGE_CC,AFFY,EVA-GONL,SANFORD_IMAGENETICS,EVA_UK10K_ALSPAC,SGDP_PRJ,COMPLETE_GENOMICS,TOPMED,CSHL,ENSEMBL,JMKIDD_LAB,EVA_GENOME_DK,EVA_MGP,GMI,EVA,HUGCELL_USP,KRIBB_YJKIM,SSMP,ABI,WEILL_CORNELL_DGM,PERLEGEN,HUMAN_LONGEVITY,SWEGEN,BCM_SSAHASNP,USC_VALOUEV,KHV_HUMAN_GENOMES,SC_JCM,KWOK,ILLUMINA,JJLAB,EVA_SVP,FSA-LAB</ns0:HANDLE>\n",
      "    <ns0:SPDI>NC_000006.12:160143443:C:G</ns0:SPDI>\n",
      "    <ns0:FXN_CLASS>intron_variant</ns0:FXN_CLASS>\n",
      "    <ns0:VALIDATED>by-frequency,by-alfa,by-cluster</ns0:VALIDATED>\n",
      "    <ns0:DOCSUM>HGVS=NC_000006.12:g.160143444C&gt;G,NC_000006.11:g.160564476C&gt;G|SEQ=[C/G]|LEN=1|GENE=SLC22A1:6580</ns0:DOCSUM>\n",
      "    <ns0:TAX_ID>9606</ns0:TAX_ID>\n",
      "    <ns0:ORIG_BUILD>83</ns0:ORIG_BUILD>\n",
      "    <ns0:UPD_BUILD>156</ns0:UPD_BUILD>\n",
      "    <ns0:CREATEDATE>2000/08/11 14:20</ns0:CREATEDATE>\n",
      "    <ns0:UPDATEDATE>2022/10/13 11:10</ns0:UPDATEDATE>\n",
      "    <ns0:SS>835338,959497,1035197,1035759,2051156,10340406,24508718,44732279,68371153,68998600,76746537,76887987,93593714,104823350,111217410,144404952,162963366,222794264,233767468,285540024,410858957,559667618,654010923,983728334,1074206279,1323142041,1581983847,1593305004,1617166610,1660160643,1711148805,1712916761,1804770792,1926970596,1946197167,1958968850,1970575644,2024151598,2152344632,2290196722,2634539291,2634539292,2635165219,2707983326,2711097223,2847573857,2986020641,3000296892,3022687346,3025866686,3347328367,3625917861,3644930307,3653213070,3668237063,3718705650,3726403168,3734174646,3744282087,3744563611,3765937827,3771337836,3809100481,3838636043,3844086157,3866169000,3913159523,3984354004,3984354005,3984448642,3984580208,3985266167,3986365874,4729251304,5237411306,5271114436,5370401956,5468490195,5558433287,5624162227,5624648564,5641929939,5843288653,5848671840,5886836419,5970986753,5979809795</ns0:SS>\n",
      "    <ns0:ALLELE>S</ns0:ALLELE>\n",
      "    <ns0:SNP_CLASS>snv</ns0:SNP_CLASS>\n",
      "    <ns0:CHRPOS>6:160143444</ns0:CHRPOS>\n",
      "    <ns0:CHRPOS_PREV_ASSM>6:160564476</ns0:CHRPOS_PREV_ASSM>\n",
      "    <ns0:TEXT/>\n",
      "    <ns0:SNP_ID_SORT>0000662138</ns0:SNP_ID_SORT>\n",
      "    <ns0:CLINICAL_SORT>0</ns0:CLINICAL_SORT>\n",
      "    <ns0:CITED_SORT/>\n",
      "    <ns0:CHRPOS_SORT>0160143444</ns0:CHRPOS_SORT>\n",
      "    <ns0:MERGED_SORT>0</ns0:MERGED_SORT>\n",
      "</ns0:DocumentSummary>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(xml_to_string(ret_snp.data[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "ret_snp_xml = ET.fromstring(xml_to_string(ret_snp.data[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Study: 1000Genomes, Frequency: G=0.121487/608\n",
      "Study: ALSPAC, Frequency: G=0.193046/744\n",
      "Study: Estonian, Frequency: G=0.154464/692\n",
      "Study: GENOME_DK, Frequency: G=0.225/9\n",
      "Study: GnomAD, Frequency: G=0.138558/19420\n",
      "Study: GoNL, Frequency: G=0.162325/162\n",
      "Study: HapMap, Frequency: G=0.117647/192\n",
      "Study: KOREAN, Frequency: G=0.001027/3\n",
      "Study: MGP, Frequency: G=0.136704/73\n",
      "Study: NorthernSweden, Frequency: G=0.155/93\n",
      "Study: PAGE_STUDY, Frequency: G=0.117242/9226\n",
      "Study: PRJEB36033, Frequency: G=0.088889/8\n",
      "Study: PRJEB37584, Frequency: G=0.002525/2\n",
      "Study: PRJEB37766, Frequency: G=0.291363/958\n",
      "Study: Qatari, Frequency: G=0.097222/21\n",
      "Study: SGDP_PRJ, Frequency: C=0.47541/58\n",
      "Study: Siberian, Frequency: C=0.444444/8\n",
      "Study: TOPMED, Frequency: G=0.141025/37328\n",
      "Study: TWINSUK, Frequency: G=0.171521/636\n",
      "Study: ALFA, Frequency: G=0.160454/13541\n",
      "Gene Name: SLC22A1\n",
      "Gene ID: 6580\n"
     ]
    }
   ],
   "source": [
    "maf_elements = ret_snp_xml.findall('.//{https://www.ncbi.nlm.nih.gov/SNP/docsum}MAF')\n",
    "for maf in maf_elements:\n",
    "    study = maf.find('{https://www.ncbi.nlm.nih.gov/SNP/docsum}STUDY').text\n",
    "    freq = maf.find('{https://www.ncbi.nlm.nih.gov/SNP/docsum}FREQ').text\n",
    "    print(f\"Study: {study}, Frequency: {freq}\")\n",
    "\n",
    "genes_element = ret_snp_xml.find('.//{https://www.ncbi.nlm.nih.gov/SNP/docsum}GENES')\n",
    "if genes_element is not None:\n",
    "    gene_name = genes_element.find('{https://www.ncbi.nlm.nih.gov/SNP/docsum}GENE_E/{https://www.ncbi.nlm.nih.gov/SNP/docsum}NAME').text\n",
    "    gene_id = genes_element.find('{https://www.ncbi.nlm.nih.gov/SNP/docsum}GENE_E/{https://www.ncbi.nlm.nih.gov/SNP/docsum}GENE_ID').text\n",
    "    \n",
    "    print(f\"Gene Name: {gene_name}\")\n",
    "    print(f\"Gene ID: {gene_id}\")\n",
    "else:\n",
    "    print(\"No GENES element found in the XML.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
